### PR TITLE
Lh/extension refactor

### LINKED ISSUE #22: Error if trying to register existing variable name for model
IndexedVarArray checks for existing indices. Also checking if a variable name is in use in a model would be a good idea before the macro registers it (and would probably have prevented the recent benchmarking bug).

### PR DESCRIPTION
**Pull request by hellemo**
Support extending standards JuMP macros (#19):

```
m = Model()
@variable(m, x[i = 1:3, j = 100:102] >= 0, container = IndexedVarArray)
```
Should also fix #22 for `IndexedVarArray`

This PR includes changes from PR #24, can separate later if we don't want those changes, but is using e.g. `nnz` here.

### CODE REVIEW

**Comment by trulsf** [comment]
Nice solution to get integration with JuMP macros. 

A question regarding the index sets. Currently, these are not copied upon construction so they can be modified during the lifetime of the variable. E.g. the following example will throw an exception
```
F = ["A", "B"]
G = [1,2,3]

@variable(m, x[F,G] >= 0, container = IndexedVarArray)

G = []
insertvar!(x, "A", 1)
```
Should it be an option to fix the index sets upon construction? I.e. let the index_names tuple have copies of the original index sets. I guess it is not that easy to provide an option to the `@variable` macro, so maybe a parametric type with the behaviour as parameter?

File: src/indexedarray.jl
Comment (trulsf) [inline src/indexedarray.jl:250]: A bit confusing to use the same index name as the function argument.

File: src/indexedarray.jl
Comment (trulsf) [inline src/indexedarray.jl:265]: I seem to get to the fallback function when using the container=IndexedVarArray, i.e. I always get the default index names.

Comment (hellemo) [reply src/indexedarray.jl:265]: Yes, you need this JuMP PR: https://github.com/jump-dev/JuMP.jl/pull/3088

**Review by trulsf — APPROVED** [review]

**Comment by hellemo** [comment]
Waiting for a decision on order of arguments (see https://github.com/jump-dev/JuMP.jl/pull/3088) before merging.

**Comment by trulsf** [comment]
I think there is a problem with the linear indexing when integers are used for the index sets
```
@variable(m, x[10:20]; container=IndexedVarArray)
x[1]
```
will lead to an infinite loop.

**Comment by trulsf** [comment]
I also have some problems with `sum(x)` and `collect(x)` where x is an `IndexedVarArray`

**Comment by trulsf** [comment]
I am leaning towards looking more to `SparseAxisArray` and not so much to `SparseArrays` for how to implement `length()` and exclude `size()`.

**Comment by hellemo** [comment]
Allright, I'll have a look, maybe revert the AbstractArray stuff if it only causes problems.

**Pull request merged**
